### ISSUE TITLE
WaveSense: estimate the wave PSF's delay, scale and offset from data

### ISSUE THREAD

**Issue by mcencini** [mcencini/bartorch#56]
`WaveSense` makes its point-spread function from the gradient wave: `max_grad`, `max_slew`, `cycles`, `adc`, `resolution`, plus `offset`, `delay` and `scale` (6d6a217). With `offset`, `delay` and `scale` at their defaults, the function is BART's `wavepsf`. Nothing in bartorch estimates those three corrections from data yet. BART doesn't either: `wavepsf`, `wave` and `wshfl` take the function as given.

This issue tracks adding that estimation as a method on the operator. `A.calibrate(kspace, ...)` would return the encoding rebuilt with the fitted wave, and expose the fitted values for reuse.

## Method to reproduce

Wave-shuffling's `shflpsf` (`examples/wave-shuffling/lib/{shflpsf,globalpsf,offsetpsf}.m`) runs on a reconstruction made with the theoretical function:

1. Take the middle slice across each phase-encode axis, zero-filled along the readout.
2. Deconvolve each slice along the readout by the ratio of a trial function to the one used for the reconstruction.
3. Score the result by the L1 norm of its readout edges, passed through a sigmoid.
4. Minimise with `fminsearch`: first the waveform's dominant Fourier coefficient per axis (amplitude and phase, which are `scale` and `delay`), then the isocentre offset, alternating.

In `script.m` this sits between an initial `wshfl` reconstruction and the final one.

## What was tried

A draft (the patch below) implements this on the native `WaveSense`: CG reconstruction, the per-axis PSF factors on the two centre slices, and a Nelder-Mead search in scaled units. It was probed on a 24³ Shepp-Logan phantom with 4 coils, a 3× oversampled readout, and wave settings `max_grad=2.7`, `max_slew=18700`, `cycles=5`, `adc=5`, `resolution=0.25`:

- **The score is right.** On slices blurred exactly by `conj(w0)·w_true` along the readout, the true settings score lowest (6.98 against 14.72 at nominal).
- **Reconstructions carry no such blur.** CG with the nominal function recovers the true image almost exactly: correlation 0.995 for errors of 2 µs, 0.5 % and 0.2 mm. The mismatch goes into the data residual instead.
- **Both scores prefer nominal.** Sharpness and data residual of the fixed reconstruction both favour the nominal settings, on fully sampled data and on 4× undersampled data.
- **Small errors are already large.** A 5 µs delay, 2 % scale and 1 mm offset put the nominal operator 139 % off the data, mostly because of the offset.

So the method could not be validated on a small synthetic phantom.

## Next step

Validate on the data the method was designed for: wave-shuffling's MPRAGE (256³, `wx = 6·256`, 16 coils; `examples/wave-shuffling/data/download.sh`). Its own `script.m` calibrates the theoretical function before the final reconstruction. If the sharpness objective works at that scale, finish the patch with tests pinned against `shflpsf`'s results on that data. If it does not, consider a joint self-calibration that alternates reconstruction with a data-consistency fit of `delay`, `scale` and `offset`.

<details>
<summary>Draft implementation (patch against 6d6a217, not merged)</summary>

```diff
diff --git a/src/bartorch/linop/mri.py b/src/bartorch/linop/mri.py
index 8a15def..b47ef79 100644
--- a/src/bartorch/linop/mri.py
+++ b/src/bartorch/linop/mri.py
@@ -288,10 +288,15 @@ class _WaveNative(_GridSense):
         basis,
         centred,
         toeplitz,
+        wave=None,
+        rebuild=None,
         **kwargs,
     ):
         from bartorch.linop.sense import _grid_ndim
 
+        self._wave = wave
+        self._rebuild = rebuild
+
         image_shape = tuple(image_shape)
         self._wave_readout = int(readout)
         if self._wave_readout < image_shape[-1]:
@@ -333,6 +338,167 @@ class _WaveNative(_GridSense):
 
         super().__init__(sensitivities, image_shape, **kwargs)
 
+    @property
+    def gradient(self) -> dict | None:
+        """The gradient wave the point-spread function was made from, or ``None``.
+
+        ``max_grad``, ``max_slew``, ``cycles``, ``adc``, ``resolution``,
+        ``offset``, ``delay`` and ``scale``, as :func:`WaveSense` took them;
+        ``None`` for a function given as ``psf=``.
+        """
+        return None if self._wave is None else dict(self._wave)
+
+    def _readout_transform(self, x: torch.Tensor, inverse: bool = False) -> torch.Tensor:
+        """The transform along the readout in this encoding's convention, unitary."""
+        fn = torch.fft.ifft if inverse else torch.fft.fft
+        if self._wave_centred:
+            x = torch.fft.ifftshift(x, dim=-1)
+            return torch.fft.fftshift(fn(x, dim=-1, norm="ortho"), dim=-1)
+        return fn(x, dim=-1, norm="ortho")
+
+    def calibrate(
+        self,
+        kspace: torch.Tensor,
+        *,
+        iterations: int = 30,
+        image: torch.Tensor | None = None,
+        sharpness: float = 100.0,
+        rounds: int = 1,
+        tol: float = 1e-6,
+    ) -> LinearOperator:
+        """This encoding with its gradient wave's delay, scale and offset fitted to ``kspace``.
+
+        Wave-shuffling's ``shflpsf``: an image reconstructed with the
+        point-spread function this encoding has, then, along the readout of
+        its middle slice across each phase-encode axis, a trial function
+        deconvolved against this one and scored by how sharp the edges it
+        leaves are.  The delays and scales and then the offsets are found by a
+        simplex search, ``rounds`` times, and the encoding is made again with
+        what was found; its :attr:`gradient` says what that was.
+
+        Only an encoding made from the gradient wave has a delay, a scale and
+        an offset to fit.
+
+        Parameters
+        ----------
+        kspace : tensor
+            The samples, as this encoding's codomain.
+        iterations : int
+            CG iterations of the reconstruction.
+        image : tensor, optional
+            A reconstruction with this encoding to fit against instead.
+        sharpness : float
+            The gain of the sigmoid the edges go through before they are
+            summed; zero sums their magnitudes.
+        rounds : int
+            How many times the delays and scales, then the offsets, are found.
+        tol : float
+            The simplex search's tolerance, in units of 0.01 ms of delay,
+            0.01 of scale and 0.1 cm of offset.
+
+        Returns
+        -------
+        LinearOperator
+            The encoding :func:`WaveSense` makes with the fitted wave.
+        """
+        import numpy as np
+        from scipy.optimize import minimize
+
+        from bartorch.optim import CG
+
+        if self._wave is None:
+            raise ValueError(
+                "only an encoding made from the gradient wave has a delay, a scale and an "
+                "offset to fit; this one was given its point-spread function"
+            )
+        if image is None:
+            image = CG(maxiter=iterations)(kspace, self)
+
+        axes = self.ndim - 1
+        spatial = tuple(self.spatial)
+        encodes, sx, wx = spatial[:-1], spatial[-1], self._wave_readout
+        wave = dict(self._wave)
+        gradient = {name: wave[name] for name in ("max_grad", "max_slew", "cycles", "adc")}
+        resolution = _per_axis(wave["resolution"], axes, "resolution")
+        values = {
+            name: list(_per_axis(wave[name], axes, name)) for name in ("delay", "scale", "offset")
+        }
+
+        # The image zero-filled along the readout as the encoding does it, and
+        # its middle slice across each phase-encode axis, each of unit norm.
+        x = torch.as_tensor(image).to(torch.complex128).reshape(-1, *spatial)
+        grown = torch.zeros(*x.shape[:-1], wx, dtype=torch.complex128)
+        start = abs(wx // 2 - sx // 2)
+        grown[..., start : start + sx] = x
+        middle = [n // 2 for n in encodes]
+        slices = [grown[:, middle[0], :, :], grown[:, :, middle[1], :]] if axes == 2 else [grown]
+        slices = [t / t.norm(dim=(-2, -1), keepdim=True).clamp_min(1e-30) for t in slices]
+        spectra = [self._readout_transform(t) for t in slices]
+
+        def slice_functions(trial):
+            factors = [
+                _wave_axis(
+                    wx,
+                    encodes[a],
+                    resolution[a],
+                    trial["offset"][a],
+                    axes == 2 and a == 0,
+                    trial["delay"][a],
+                    trial["scale"][a],
+                    **gradient,
+                )
+                for a in range(axes)
+            ]
+            if axes == 2:
+                fz, fy = factors
+                return [fz[middle[0]][None, :] * fy, fz * fy[middle[1]][None, :]]
+            return factors
+
+        made_with = slice_functions(values)
+
+        def score(trial):
+            total = 0.0
+            for spectrum, w0, wt in zip(spectra, made_with, slice_functions(trial)):
+                deblurred = self._readout_transform(spectrum * (wt.conj() * w0), inverse=True)
+                edge = (deblurred - torch.roll(deblurred, 1, dims=-1)).abs()
+                if sharpness > 0:
+                    edge = torch.sigmoid(sharpness * edge) - 0.5
+                total += float(edge.sum())
+            return total
+
+        units = {"delay": 0.01, "scale": 0.01, "offset": 0.1}
+
+        def search(names):
+            begin = [values[name][a] for name in names for a in range(axes)]
+            layout = [(name, a) for name in names for a in range(axes)]
+
+            def trial_of(u):
+                trial = {name: list(v) for name, v in values.items()}
+                for (name, a), b, step in zip(layout, begin, u):
+                    trial[name][a] = b + float(step) * units[name]
+                return trial
+
+            count = len(begin)
+            found = minimize(
+                lambda u: score(trial_of(u)),
+                np.zeros(count),
+                method="Nelder-Mead",
+                options={
+                    "initial_simplex": np.vstack([np.zeros(count), np.eye(count)]),
+                    "xatol": tol,
+                    "fatol": tol,
+                    "maxiter": 400 * count,
+                },
+            )
+            values.update(trial_of(found.x))
+
+        for _ in range(rounds):
+            search(("delay", "scale"))
+            search(("offset",))
+
+        fitted = {name: (tuple(v) if axes == 2 else v[0]) for name, v in values.items()}
+        return WaveSense(**{**self._rebuild, **fitted})
+
     def _grid_encoding(self):
         return () if self._grid_basis is None else (int(self._grid_basis.shape[1]),)
 
@@ -585,6 +751,13 @@ def _per_axis(value, n: int, what: str) -> tuple[float, ...]:
     return values
 
 
+def _wave_axis(readout, n, step, shift, cosine, delay, scale, max_grad, max_slew, cycles, adc):
+    """One axis's factor of the wave point-spread function, ``(n, readout)``."""
+    ppcm = _wave_phase_per_cm(readout, cycles, max_grad, max_slew, adc, cosine, delay, scale)
+    location = step * (torch.arange(n, dtype=torch.float64) - n // 2) - shift
+    return torch.exp(-1j * location[:, None] * ppcm[None, :])
+
+
 def _wave_psf(
     readout, encodes, max_grad, max_slew, cycles, adc, resolution, offset, delay=0.0, scale=1.0
 ) -> torch.Tensor:
@@ -597,21 +770,33 @@ def _wave_psf(
     :func:`_wave_phase_per_cm` applies them.
     """
     encodes = tuple(int(n) for n in encodes)
-    resolutions = _per_axis(resolution, len(encodes), "resolution")
-    offsets = _per_axis(offset, len(encodes), "offset")
-    delays = _per_axis(delay, len(encodes), "delay")
-    scales = _per_axis(scale, len(encodes), "scale")
-
+    per = {
+        name: _per_axis(value, len(encodes), name)
+        for name, value in (
+            ("resolution", resolution),
+            ("offset", offset),
+            ("delay", delay),
+            ("scale", scale),
+        )
+    }
     out = torch.ones((*encodes, readout), dtype=torch.complex128)
-    for axis, (n, step, shift) in enumerate(zip(encodes, resolutions, offsets)):
-        cosine = len(encodes) == 2 and axis == 0
-        ppcm = _wave_phase_per_cm(
-            readout, cycles, max_grad, max_slew, adc, cosine, delays[axis], scales[axis]
+    for axis, n in enumerate(encodes):
+        factor = _wave_axis(
+            readout,
+            n,
+            per["resolution"][axis],
+            per["offset"][axis],
+            len(encodes) == 2 and axis == 0,
+            per["delay"][axis],
+            per["scale"][axis],
+            max_grad,
+            max_slew,
+            cycles,
+            adc,
         )
-        location = step * (torch.arange(n, dtype=torch.float64) - n // 2) - shift
         shape = [1] * len(encodes) + [readout]
         shape[axis] = n
-        out = out * torch.exp(-1j * location[:, None] * ppcm[None, :]).reshape(shape)
+        out = out * factor.reshape(shape)
     return out
 
 
@@ -901,7 +1086,27 @@ def WaveSense(  # noqa: N802  (it is a constructor)
         "adc": adc,
         "resolution": resolution,
     }
+    rebuild = dict(
+        sensitivities=sensitivities,
+        image_shape=image_shape,
+        readout=readout,
+        pattern=pattern,
+        centred=centred,
+        **wave,
+        offset=offset,
+        delay=delay,
+        scale=scale,
+        positions=positions,
+        basis=basis,
+        toeplitz=toeplitz,
+        kernels=kernels,
+        coil_batch=coil_batch,
+        device=device,
+        ndim=ndim,
+    )
+    made_from = None
     if psf is None:
+        made_from = {**wave, "offset": offset, "delay": delay, "scale": scale}
         missing = [name for name, value in wave.items() if value is None]
         if missing:
             raise ValueError(f"give psf=, or the gradient wave: {', '.join(missing)} missing")
@@ -932,6 +1137,8 @@ def WaveSense(  # noqa: N802  (it is a constructor)
         coil_batch=coil_batch,
         device=device,
         ndim=ndim,
+        wave=made_from,
+        rebuild=rebuild,
     )
 
 
```

</details>
